password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,222 @@
+{-# LANGUAGE CPP               #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+
+{-|
+Module      : Data.Password.Valid
+Copyright   : (c) Dennis Gosnell, 2019; Felix Paulusma, 2020; Hiroto Shioi, 2020
+License     : BSD-style (see LICENSE file)
+Maintainer  : cdep.illabout@gmail.com
+Stability   : experimental
+Portability : POSIX
+
+= validate
+
+It is common for passwords to have a set of requirements.
+(e.g. Password must contain at least eight characters that consist of alphabetic
+characters combined with numbers or special characters.)
+
+Validate module provides set of functions which enables you to validate them.
+
+-}
+
+module Data.Password.Validate
+  ( -- * Data type
+    PasswordPolicy (..),
+    InvalidReason (..),
+    CharacterCategory(..),
+    -- * Default
+    defaultPasswordPolicy,
+    -- * Functions
+    isValidPassword,
+    validatePassword,
+    -- * Utility
+    isValidPasswordPolicy,
+    defaultCharSet,
+  ) where
+
+import Data.Char (isDigit, isLower, isUpper)
+import Data.Maybe (catMaybes, fromMaybe)
+import Data.Password.Internal (Password (..))
+#if! MIN_VERSION_base(4,13,0)
+import Data.Semigroup ((<>))
+#endif
+import Data.Text (Text)
+import qualified Data.Text as T
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+
+-- | Set of policies used to validate 'Password'
+data PasswordPolicy = PasswordPolicy
+    { minimumLength  :: !Int
+    -- ^ Required password minimum length
+    , maximumLength  :: !Int
+    -- ^ Required password maximum length
+    , uppercaseChars :: !(Maybe Int)
+    -- ^ Required number of upper-case characters
+    , lowercaseChars :: !(Maybe Int)
+    -- ^ Required number of lower-case characters
+    , specialChars   :: !(Maybe Int)
+    -- ^ Required number of special characters
+    , digitChars     :: !(Maybe Int)
+    -- ^ Required number of ASCII-digit characters
+    , charSet        :: !Text
+    -- ^ Set of characters that can be used for password
+    }
+    deriving (Eq, Ord, Show)
+
+-- | Default value for the 'PasswordPolicy'
+defaultPasswordPolicy :: PasswordPolicy
+defaultPasswordPolicy = PasswordPolicy
+  { minimumLength = 8,
+    maximumLength = 32,
+    uppercaseChars = Just 1,
+    lowercaseChars = Just 1,
+    specialChars = Nothing,
+    digitChars = Just 1,
+    charSet = defaultCharSet
+  }
+
+-- | Default character sets consist of uppercase, lowercase letters, numbers,
+-- and special characters
+defaultCharSet :: Text
+defaultCharSet = T.pack $ specialLetters <> ['A' .. 'Z'] <> ['a' .. 'z'] <> ['0' .. '9']
```

Comment:
`defaultCharSet` can also be written as `T.pack [' ' .. '~']` which is the entire ASCII set without the control characters.

`specialLetters` I'd maybe call `specialCharacters` and define as
```hs
import Data.Char as C

specialCharacters :: String
specialCharacters = filter (\c -> not $ any [C.isDigit c, C.isAsciiUpper c, C.isAsciiLower c]) defaultCharSet
```
Just to have it conceptually obvious. Or alternatively, write a unit test that makes sure that it equates the current `specialLetters` list to the one I provided. That way we can be sure we didn't miss any special characters or included incorrect ones.

It's also a good idea to make some unit/regression tests to make sure the defaultCharSet is the Ascii set without control characters.
* defaultSet contains no control characters
* all characters in defaultSet are all `ord c < 126`